model/waterfall.go

```diff
@@ -0,0 +1,168 @@
+		},
+	}
+
+	// TODO DEVPROD-10177: Add revision order logic to handle pagination.
+
+	pipeline := []bson.M{{"$match": match}}
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
+	limit := defaultVersionLimit
+	if opts.Limit != 0 {
+		limit = opts.Limit
+	}
+
+	pipeline = append(pipeline, bson.M{"$limit": limit})
+
+	res := []Version{}
+	env := evergreen.GetEnvironment()
+	cursor, err := env.DB().Collection(VersionCollection).Aggregate(ctx, pipeline)
+	if err != nil {
+		return nil, errors.Wrap(err, "aggregating versions")
+	}
+	err = cursor.All(ctx, &res)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// GetWaterfallBuildVariants returns all build variants associated with the specified versions. Each build variant contains an array of builds sorted by revision and their tasks.
+func GetWaterfallBuildVariants(ctx context.Context, versions []Version) ([]WaterfallBuildVariant, error) {
+	versionIds := []string{}
+	for _, version := range versions {
+		versionIds = append(versionIds, version.Id)
+	}
+
+	pipeline := []bson.M{{"$match": bson.M{VersionIdKey: bson.M{"$in": versionIds}}}}
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
+
+	pipeline = append(pipeline, bson.M{"$unwind": bson.M{"path": "$" + VersionBuildVariantsKey}})
+	buildsKey := "builds"
+	pipeline = append(pipeline, bson.M{
+		"$group": bson.M{
+			"_id": "$" + bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusVariantKey),
+			buildsKey: bson.M{
+				"$push": "$" + bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusIdKey),
+			},
+		},
+	})
+	pipeline = append(pipeline, bson.M{
+		// TODO DEVPROD-10178: Should be able to filter on build variant ID/name here.
+		"$lookup": bson.M{
+			"from":         build.Collection,
+			"localField":   buildsKey,
+			"foreignField": build.IdKey,
+			"as":           buildsKey,
+		},
+	})
+	pipeline = append(pipeline, bson.M{"$unwind": bson.M{"path": "$" + buildsKey}})
+
+	// Join all tasks that appear in the build's task cache and overwrite the list of task IDs with partial task documents
+	pipeline = append(pipeline, bson.M{"$lookup": bson.M{
+		"from":         task.Collection,
+		"localField":   bsonutil.GetDottedKeyName(buildsKey, build.TasksKey, build.TaskCacheIdKey),
+		"foreignField": task.IdKey,
+		"pipeline": []bson.M{
+			{
+				"$sort": bson.M{task.IdKey: 1},
+			},
+			// TODO DEVPROD-10179, DEVPROD-10180: Should be able to filter on the returned task names and statuses here.
+			{
+				"$project": bson.M{
+					task.IdKey:          1,
+					task.StatusKey:      1,
+					task.DisplayNameKey: 1,
+				},
+			},
+		},
+		"as": bsonutil.GetDottedKeyName(buildsKey, build.TasksKey),
+	},
+	})
+	// Sorting builds here guarantees a consistent order in the subsequent $group stage
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{"_id": 1, bsonutil.GetDottedKeyName(buildsKey, build.RevisionOrderNumberKey): -1}})
```

Comment:
Maybe I'm getting a tad lost in the aggregation pipeline 😅 why does this  have to sort first by `_id`, and then by revision order number? Is it to get the list of build variants in order by name? I thought that the sort on L146 is already accomplishing that. If it's intentional, maybe adding a small comment to the follow-up sort would help to explain why we have to sort on `_id` twice.

Reply:
This ID sort is superfluous 🙏 great catch!